scandium-core/src/main/java/org/eclipse/californium/scandium/CookieGenerator.java

```diff
@@ -0,0 +1,125 @@
+/*******************************************************************************
+ * Copyright (c) 2017 Institute for Pervasive Computing, ETH Zurich and others.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ *
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ *
+ */
+package org.eclipse.californium.scandium;
+
+import java.security.GeneralSecurityException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
+
+import org.eclipse.californium.scandium.dtls.ClientHello;
+import org.eclipse.californium.scandium.dtls.CompressionMethod;
+import org.eclipse.californium.scandium.dtls.cipher.CipherSuite;
+
+/**
+ * Generates a cookie in such a way that they can be verified without retaining
+ * any per-client state on the server.
+ *
+ * <pre>
+ * Cookie = HMAC(Secret, Client - IP, Client - Parameters)
+ * </pre>
+ *
+ * as suggested
+ * <a href="http://tools.ietf.org/html/rfc6347#section-4.2.1">here</a>.
+ *
+ */
+public class CookieGenerator {
+
+	private static final long keyLifetime = TimeUnit.MINUTES.toMillis(5);
+
+	private long lastGenerationDate;
+	private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+	private Mac hmac;
```

Comment:
I'm not sure to see how I can use it as I must change `hmac` and `lastGenerationDate` in an atomic way.